src/mesh_smoothing.rs

```diff
@@ -91,6 +94,267 @@ pub fn laplacian_smoothing(
     )
 }
 
+/// Performs one iteration of Loop Subdivision on geometry.
+///
+/// The subdivision works in two steps:
+///
+/// 1) Split each triangle into 4 smaller triangles,
+/// 2) Update the position of each vertex of the mesh based on
+///    weighted averages of its neighboring vertex positions,
+///    depending on where the vertex is in the topology and whether
+///    the vertex is newly created, or did already exist.
+///
+/// The geometry **must** be triangulated.
+///
+/// Implementation based on [mdfisher's
+/// post](https://graphics.stanford.edu/~mdfisher/subdivision.html).
+pub fn loop_subdivision(
+    geometry: &Geometry,
+    vertex_to_vertex_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
+    face_to_face_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
+) -> Geometry {
+    #[derive(Debug, Eq)]
+    struct UnorderedPair(u32, u32);
+
+    impl PartialEq for UnorderedPair {
+        fn eq(&self, other: &Self) -> bool {
+            self.0 == other.0 && self.1 == other.1 || self.0 == other.1 && self.1 == other.0
+        }
+    }
+
+    impl Hash for UnorderedPair {
+        fn hash<H: Hasher>(&self, state: &mut H) {
+            cmp::min(self.0, self.1).hash(state);
+            cmp::max(self.0, self.1).hash(state);
+        }
+    }
+
+    let mut vertices: Vec<Point3<f32>> = geometry.vertices().iter().copied().collect();
+
+    // Relocate existing vertices first
+    for (i, vertex) in vertices.iter_mut().enumerate() {
+        let neighbors = &vertex_to_vertex_topology[&cast_u32(i)];
+
+        match neighbors.len() {
+            // N == 0 means this is an orphan vertex. N == 1 can't
+            // happen in our mesh representation.
+            0 | 1 => (),
+            2 => {
+                // N == 2 means this is a naked edge vertex.
+                // Use (3/4, 1/8, 1/8) relocation scheme.
+
+                let vi1 = cast_usize(neighbors[0]);
+                let vi2 = cast_usize(neighbors[1]);
+
+                let v0 = geometry.vertices()[i];
+                let v1 = geometry.vertices()[vi1];
+                let v2 = geometry.vertices()[vi2];
+
+                *vertex = Point3::origin()
+                    + v0.coords * 3.0 / 4.0
+                    + v1.coords * 1.0 / 8.0
+                    + v2.coords * 1.0 / 8.0;
+            }
+            3 => {
+                // N == 3 means the vertex is not a part of a naked edge.
```

Comment:
Why not? It might.

Reply:
Damn... my source article did not cover this case or provide a relocation scheme for it. Do we want handle that case somewhat specially, or should I just update the comment?

Reply:
If your algorithm cares about edge valency, then it's good as it is and I suggest changing just the comments. But if you actually need to identify vertices on the borders, then please use one of the existing functions doing exactly that. 

A side note: it might be useful to have the option to preserve the edges during smoothing. It's not that much extra work and we have the same functionality for the Laplacian smoothing.

Reply:
> If your algorithm cares about edge valency, then it's good as it is and I suggest changing just the comments.

Naked edges were not specified in sources I read. It only cares about valency.

Reply:
> A side note: it might be useful to have the option to preserve the edges during smoothing. It's not that much extra work and we have the same functionality for the Laplacian smoothing.

I think this should be do-able. It'll just skip the `N == 2` branch if this is enabled. I'll add it.

Reply:
Hmmm, wait. What exactly do you mean by that? Preserve naked edges (not relocate vertices that are part of a naked edge)? Or all edges?

Reply:
> > A side note: it might be useful to have the option to preserve the edges during smoothing. It's not that much extra work and we have the same functionality for the Laplacian smoothing.
> 
> I think this should be do-able. It'll just skip the `N == 2` branch if this is enabled. I'll add it.

that's not enough - it only preserves "corners". you'll need to check for real border vertices. but we can keep it for later

Reply:
> Hmmm, wait. What exactly do you mean by that? Preserve naked edges (not relocate vertices that are part of a naked edge)? Or all edges?

preserve naked edges, including the newly created vertices in the middle of the existing naked edges.
but as i said, this might be noted as a new feature and implemented later

Reply:
Ok, I'd keep it for later then.